### PR TITLE
add topic filter to get articles endpoint

### PR DESCRIPTION
**Pull request by nldblanch**

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -243,87 +243,131 @@ describe("/api/articles", () => {
...
+          return request(app)
+            .get("/api/articles?sort_by=author&&order=anything_I_want")
+            .expect(200)
+            .then(({ body: { articles } }) => {
+              expect(articles).toBeSortedBy("author", {
+                descending: false,
+              });
+            });
+        });
+      })
```
Comment (Harrison365) [inline __tests__/app.test.js:328]: ^^^ I'm assuming you had these checked in your ticket 11 PR??? ^^^

File: __tests__/app.test.js [hunk 2]
```diff
@@ -243,87 +243,131 @@ describe("/api/articles", () => {
...
+          const input_topic = "mitch"
+          return request(app)
+          .get(`/api/articles?topic=${input_topic}`)
           .expect(200)
           .then(({ body: { articles } }) => {
-            expect(articles).toBeSortedBy("votes", {
-              descending: true,
-            });
+            articles.forEach((article) => {
+              expect(article.topic).toBe(input_topic)
```
Comment (Harrison365) [inline __tests__/app.test.js:337]: Don't forget a length check here :)

File: __tests__/app.test.js [hunk 3]
```diff
@@ -243,87 +243,131 @@ describe("/api/articles", () => {
...
-          .expect(200)
-          .then(({ body: { articles } }) => {
-            expect(articles).toBeSortedBy("votes", {
-              descending: false,
-            });
+            expect(message).toEqual("404 - Not Found")
           });
-      });
-      it("200: defaults to descending (alphabetical ascending) when invalid order given", () => {
+        })
```
Comment (Harrison365) [inline __tests__/app.test.js:358]: Great 👍

File: models/models.js [hunk 4]
```diff
@@ -29,29 +29,45 @@ exports.fetchArticleById = (id) => {
...
-  ORDER BY articles.%I
+  ON articles.article_id = comments.article_id
   `;
+  let dataQueries = []
+
+  if (topic) {
+    promiseArray.push(checkExists("topics", "slug", topic))
+    stringQuery += ` WHERE topic = $1`
+    dataQueries.push(topic)
+  }
```
Comment (Harrison365) [inline models/models.js:50]: Nice 👍

File: models/models.js [hunk 5]
```diff
@@ -29,29 +29,45 @@ exports.fetchArticleById = (id) => {
...
   //descending alphabetical order is ascending numerical order
   if (sort_by === "title" || sort_by === "topic" || sort_by === "author") {
     order = "asc";
   }
-  stringQuery += order === "asc" ? "ASC" : "DESC";
+  stringQuery += order === "asc" ? " ASC" : " DESC";
 
   const formattedStringQuery = format(stringQuery, sort_by);
-  return db.query(formattedStringQuery).then(({ rows }) => {
+  promiseArray.unshift(db.query(formattedStringQuery, dataQueries))
```
Comment (Harrison365) [inline models/models.js:62]: You shouldn't need pg-format outside of seeding. $1, $2 etc. is preferable :)

File: models/models.js [hunk 6]
```diff
@@ -29,29 +29,45 @@ exports.fetchArticleById = (id) => {
...
+  return Promise.all(promiseArray)
+  .then(([{rows}, exists = {exists: true}]) => {
+    
+    if (rows.length === 0 && !exists.exists) {
+      return Promise.reject({ status: 404, message: "404 - Not Found" });
+    }
```
Comment (Harrison365) [inline models/models.js:68]: Nice work 🙂

File: models/models.js (hunk 5 above)
Comment (nldblanch) [reply models/models.js:62]: That was in there for a %I because it's ordering by a column name - I though $1 or $2 doesn't work in those cases?

**Pull request merged**